File: app/src/main/java/com/example/rooster/financials/transactions/TransactionMonitoringScreen.kt
package com.example.rooster.financials.transactions

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Flag
import androidx.compose.material.icons.filled.Search
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.update
import java.text.DecimalFormat
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import kotlin.random.Random

// --- Data Classes ---
data class ManualReviewFlag(
    val reason: String,
    val adminId: String,
    val timestamp: Date
) {
    fun getFormattedTimestamp(): String = SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.getDefault()).format(timestamp)
}

data class Transaction(
    val transactionId: String,
    val timestamp: Date,
    val userId: String,
    val amount: Double,
    val currency: String = "USD",
    val paymentMethod: String,
    val status: String, // completed, pending, failed, refunded, disputed
    val transactionType: String,
    val description: String,
    val fraudFlags: List<String>,
    val ipAddress: String,
    val originalTransactionId: String? = null,
    var manualReviewFlags: MutableList<ManualReviewFlag> = mutableListOf()
) {
    fun getFormattedTimestamp(): String = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault()).format(timestamp)
}

// --- Filter State ---
data class TransactionFilters(
    val status: String? = null,
    val userId: String? = null,
    val minAmount: Double? = null,
    val hasFraudFlags: Boolean? = null // null = any, true = has flags, false = no flags
)

// --- ViewModel ---
class TransactionMonitoringViewModel : ViewModel() {
    private val _allTransactions = MutableStateFlow<List<Transaction>>(emptyList())

    private val _filters = MutableStateFlow(TransactionFilters())
    val filters: StateFlow<TransactionFilters> = _filters

    val filteredTransactions: StateFlow<List<Transaction>> = combine(_allTransactions, _filters) { transactions, filters ->
        transactions.filter { tx ->
            (filters.status == null || tx.status.equals(filters.status, ignoreCase = true)) &&
            (filters.userId == null || tx.userId.equals(filters.userId, ignoreCase = true)) &&
            (filters.minAmount == null || tx.amount >= filters.minAmount) &&
            (filters.hasFraudFlags == null || (filters.hasFraudFlags == true && tx.fraudFlags.isNotEmpty()) || (filters.hasFraudFlags == false && tx.fraudFlags.isEmpty()))
        }.sortedByDescending { it.timestamp }
    }.stateIn(kotlinx.coroutines.MainScope(), kotlinx.coroutines.flow.SharingStarted.WhileSubscribed(5000), emptyList())


    init {
        loadMockTransactions()
    }

    private fun loadMockTransactions(count: Int = 50) {
        val paymentMethods = listOf("credit_card", "paypal", "bank_transfer", "crypto")
        val statuses = listOf("completed", "pending", "failed", "refunded", "disputed")
        val txTypes = listOf("consultation_fee", "marketplace_sale", "subscription", "payout_fee", "refund")
        val userIds = (1..5).map { "user${it.toString().padStart(3, '0')}" }
        val productIds = (1..3).map { "prod${it.toString().padStart(3, '0')}" }
        val now = System.currentTimeMillis()

        val tempTransactions = mutableListOf<Transaction>()
        for (i in 0 until count) {
            val amount = Random.nextDouble(5.0, 500.0)
            val status = statuses.random()
            val fraudFlags = mutableListOf<String>()
            if (amount > 400 && Random.nextBoolean()) fraudFlags.add("high_value_transaction")
            if (status == "failed" && Random.nextDouble() < 0.2) fraudFlags.add("multiple_failed_attempts_suspected")
            if (Random.nextDouble() < 0.05) fraudFlags.add("unusual_location_match")

            tempTransactions.add(
                Transaction(
                    transactionId = "txn_${now - Random.nextLong(0, 30 * 86400000L * 24 * 60)}_$i",
                    timestamp = Date(now - Random.nextLong(0, 30 * 86400000L)),
                    userId = userIds.random(),
                    amount = amount,
                    paymentMethod = paymentMethods.random(),
                    status = status,
                    transactionType = txTypes.random(),
                    description = "${txTypes.random().replace('_', ' ').titlecase()} for ${if (txTypes.last().contains("sale")) productIds.random() else userIds.random()}",
                    fraudFlags = fraudFlags,
                    ipAddress = "192.168.1.${Random.nextInt(1, 254)}",
                    originalTransactionId = if (status == "refunded") "txn_orig_${Random.nextInt(1000)}" else null
                )
            )
        }
        _allTransactions.value = tempTransactions
    }

    fun updateFilters(newFilters: TransactionFilters) {
        _filters.value = newFilters
    }

    fun flagTransactionForReview(transactionId: String, reason: String, adminId: String) {
        _allTransactions.update { transactions ->
            transactions.map {
                if (it.transactionId == transactionId) {
                    it.copy(manualReviewFlags = (it.manualReviewFlags + ManualReviewFlag(reason, adminId, Date())).toMutableList())
                } else it
            }
        }
    }
}
// Helper for title case if needed elsewhere
fun String.titlecase(): String = replaceFirstChar { if (it.isLowerCase()) it.titlecase(Locale.getDefault()) else it.toString() }


// --- Composable Screen ---
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun TransactionMonitoringScreen(viewModel: TransactionMonitoringViewModel = androidx.lifecycle.viewmodel.compose.viewModel()) {
    val transactions by viewModel.filteredTransactions.collectAsState()
    val currentFilters by viewModel.filters.collectAsState()

    var showFilterDialog by remember { mutableStateOf(false) }
    var showFlagDialog by remember { mutableStateOf(false) }
    var transactionToFlag by remember { mutableStateOf<Transaction?>(null) }
    var flagReason by remember { mutableStateOf("") }

    if (showFilterDialog) {
        FilterDialog(
            currentFilters = currentFilters,
            onDismiss = { showFilterDialog = false },
            onApplyFilters = { newFilters ->
                viewModel.updateFilters(newFilters)
                showFilterDialog = false
            }
        )
    }

    if (showFlagDialog && transactionToFlag != null) {
        FlagTransactionDialog(
            transaction = transactionToFlag!!,
            reason = flagReason,
            onReasonChange = { flagReason = it },
            onDismiss = { showFlagDialog = false; transactionToFlag = null; flagReason = "" },
            onConfirm = {
                viewModel.flagTransactionForReview(transactionToFlag!!.transactionId, flagReason, "admin_compose")
                showFlagDialog = false
                transactionToFlag = null
                flagReason = ""
            }
        )
    }


    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Transaction Monitoring") },
                colors = TopAppBarDefaults.topAppBarColors(containerColor = Color(0xFF78909C)), // Blue Grey
                actions = {
                    IconButton(onClick = { showFilterDialog = true }) {
                        Icon(Icons.Filled.Search, contentDescription = "Filter Transactions")
                    }
                }
            )
        }
    ) { paddingValues ->
        LazyColumn(modifier = Modifier.padding(paddingValues).padding(horizontal = 16.dp, vertical = 8.dp)) {
            if (transactions.isEmpty()){
                item {
                    Box(modifier = Modifier.fillParentMaxSize(), contentAlignment = Alignment.Center) {
                        Text("No transactions match current filters.")
                    }
                }
            } else {
                items(transactions, key = { it.transactionId }) { tx ->
                    TransactionCard(tx) {
                        transactionToFlag = tx
                        showFlagDialog = true
                    }
                }
            }
        }
    }
}

@Composable
fun TransactionCard(tx: Transaction, onFlagClick: () -> Unit) {
    val df = remember { DecimalFormat("$ #,##0.00") }
    Card(modifier = Modifier.fillMaxWidth().padding(vertical = 6.dp), elevation = CardDefaults.cardElevation(2.dp)) {
        Column(Modifier.padding(12.dp)) {
            Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween){
                Text(tx.transactionId, style = MaterialTheme.typography.titleSmall)
                IconButton(onClick = onFlagClick, modifier = Modifier.size(24.dp)) {
                    Icon(Icons.Filled.Flag, "Flag for Review", tint = if (tx.manualReviewFlags.isNotEmpty()) Color.Red else Color.Gray)
                }
            }
            Text("User: ${tx.userId}, Amount: ${df.format(tx.amount)} ${tx.currency}")
            Text("Type: ${tx.transactionType.titlecase()}, Method: ${tx.paymentMethod.titlecase()}")
            Text("Status: ${tx.status.titlecase()}", color = when(tx.status){
                "completed" -> Color.Green
                "pending" -> Color.Blue
                "failed" -> Color.Red
                "refunded" -> Color.Magenta
                "disputed" -> Color.Yellow // Choose appropriate color
                else -> Color.DarkGray
            })
            Text("Time: ${tx.getFormattedTimestamp()}, IP: ${tx.ipAddress}")
            if (tx.fraudFlags.isNotEmpty()) {
                Text("Fraud Flags: ${tx.fraudFlags.joinToString()}", color = MaterialTheme.colorScheme.error, fontWeight = FontWeight.Bold)
            }
            tx.manualReviewFlags.lastOrNull()?.let {
                 Text("Manual Review: ${it.reason} by ${it.adminId} at ${it.getFormattedTimestamp()}", color = Color.Red, style = MaterialTheme.typography.bodySmall)
            }
        }
    }
}

@Composable
fun FilterDialog(
    currentFilters: TransactionFilters,
    onDismiss: () -> Unit,
    onApplyFilters: (TransactionFilters) -> Unit
) {
    var status by remember { mutableStateOf(currentFilters.status ?: "") }
    var userId by remember { mutableStateOf(currentFilters.userId ?: "") }
    var minAmount by remember { mutableStateOf(currentFilters.minAmount?.toString() ?: "") }
    var hasFraudFlags by remember { mutableStateOf(currentFilters.hasFraudFlags) } // null, true, false

    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text("Filter Transactions") },
        text = {
            Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                OutlinedTextField(value = status, onValueChange = { status = it }, label = { Text("Status (e.g., completed)") })
                OutlinedTextField(value = userId, onValueChange = { userId = it }, label = { Text("User ID") })
                OutlinedTextField(value = minAmount, onValueChange = { minAmount = it.filter{c -> c.isDigit() || c == '.'} }, label = { Text("Min Amount") })
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text("Fraud Flags: ")
                    Spacer(Modifier.width(8.dp))
                    RadioButton(selected = hasFraudFlags == true, onClick = { hasFraudFlags = true })
                    Text("Yes")
                    Spacer(Modifier.width(8.dp))
                    RadioButton(selected = hasFraudFlags == false, onClick = { hasFraudFlags = false })
                    Text("No")
                     Spacer(Modifier.width(8.dp))
                    RadioButton(selected = hasFraudFlags == null, onClick = { hasFraudFlags = null })
                    Text("Any")
                }
            }
        },
        confirmButton = {
            Button(onClick = {
                onApplyFilters(TransactionFilters(
                    status = status.ifBlank { null },
                    userId = userId.ifBlank { null },
                    minAmount = minAmount.toDoubleOrNull(),
                    hasFraudFlags = hasFraudFlags
                ))
            }) { Text("Apply") }
        },
        dismissButton = { Button(onClick = onDismiss) { Text("Cancel") } }
    )
}

@Composable
fun FlagTransactionDialog(
    transaction: Transaction,
    reason: String,
    onReasonChange: (String) -> Unit,
    onDismiss: () -> Unit,
    onConfirm: () -> Unit
) {
    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text("Flag Transaction: ${transaction.transactionId}") },
        text = {
            OutlinedTextField(
                value = reason,
                onValueChange = onReasonChange,
                label = { Text("Reason for flagging") },
                modifier = Modifier.fillMaxWidth()
            )
        },
        confirmButton = { Button(onClick = onConfirm, enabled = reason.isNotBlank()) { Text("Flag") } },
        dismissButton = { Button(onClick = onDismiss) { Text("Cancel") } }
    )
}

@Preview(showBackground = true)
@Composable
fun PreviewTransactionMonitoringScreen() {
    MaterialTheme {
        TransactionMonitoringScreen(viewModel = TransactionMonitoringViewModel())
    }
}

File: app/financials/transaction_monitoring_screen.py
import datetime
import random

class TransactionMonitoringScreen:
    def __init__(self):
        self.transactions = []
        self._initialize_mock_transactions()

    def _initialize_mock_transactions(self, count=50):
        # Mock data for transactions
        payment_methods = ["credit_card", "paypal", "bank_transfer", "crypto"]
        statuses = ["completed", "pending", "failed", "refunded", "disputed"]
        transaction_types = ["consultation_fee", "marketplace_sale", "subscription", "payout_fee", "refund"]
        user_ids = [f"user{str(i).zfill(3)}" for i in range(1, 20)]
        product_ids = [f"prod{str(i).zfill(3)}" for i in range(1, 10)]

        for i in range(count):
            timestamp = datetime.datetime.now() - datetime.timedelta(days=random.randint(0, 30), hours=random.randint(0,23), minutes=random.randint(0,59))
            amount = round(random.uniform(5.0, 500.0), 2)
            status = random.choice(statuses)

            # Basic fraud flags
            fraud_flags = []
            if amount > 400 and random.random() < 0.3:
                fraud_flags.append("high_value_transaction")
            if status == "failed" and random.random() < 0.2:
                 fraud_flags.append("multiple_failed_attempts_suspected") # This would need more context in reality
            if random.random() < 0.05: # Small chance for any transaction
                fraud_flags.append("unusual_location_match") # Mock, real system would check IP/Geo

            transaction = {
                "transaction_id": f"txn_{timestamp.strftime('%Y%m%d')}_{str(i).zfill(4)}",
                "timestamp": timestamp,
                "user_id": random.choice(user_ids),
                "amount": amount,
                "currency": "USD",
                "payment_method": random.choice(payment_methods),
                "status": status,
                "transaction_type": random.choice(transaction_types),
                "description": f"{random.choice(transaction_types).replace('_', ' ').title()} for {random.choice(product_ids) if 'sale' in transaction_types[-1] else random.choice(user_ids)}",
                "fraud_flags": fraud_flags,
                "ip_address": f"192.168.1.{random.randint(1,254)}" # Mock IP
            }
            if status == "refunded":
                transaction["original_transaction_id"] = f"txn_{timestamp.strftime('%Y%m%d')}_{str(random.randint(0,i-1 if i>0 else 0)).zfill(4)}"

            self.transactions.append(transaction)

        # Sort by timestamp descending
        self.transactions.sort(key=lambda x: x["timestamp"], reverse=True)

    def get_all_transactions(self):
        return self.transactions

    def get_transaction_by_id(self, transaction_id):
        for tx in self.transactions:
            if tx["transaction_id"] == transaction_id:
                return tx
        return None

    def filter_transactions(self, status=None, user_id=None, min_amount=None, has_fraud_flags=None):
        results = self.transactions
        if status:
            results = [tx for tx in results if tx["status"] == status]
        if user_id:
            results = [tx for tx in results if tx["user_id"] == user_id]
        if min_amount is not None:
            results = [tx for tx in results if tx["amount"] >= min_amount]
        if has_fraud_flags is True:
            results = [tx for tx in results if len(tx["fraud_flags"]) > 0]
        if has_fraud_flags is False: # Explicitly filter for no flags
            results = [tx for tx in results if len(tx["fraud_flags"]) == 0]
        return results

    def flag_transaction_for_review(self, transaction_id, reason, admin_id):
        tx = self.get_transaction_by_id(transaction_id)
        if tx:
            if "manual_review_flags" not in tx:
                tx["manual_review_flags"] = []
            tx["manual_review_flags"].append({
                "reason": reason,
                "admin_id": admin_id,
                "timestamp": datetime.datetime.now()
            })
            print(f"Transaction {transaction_id} flagged for manual review by {admin_id}: {reason}")
            return True
        print(f"Transaction {transaction_id} not found.")
        return False

    def display_screen(self):
        print("---- Transaction Monitoring Screen ----")

        print("\n-- Recent Transactions (Top 5) --")
        for tx in self.get_all_transactions()[:5]:
            print(f"  ID: {tx['transaction_id']}, Time: {tx['timestamp'].strftime('%Y-%m-%d %H:%M')}, User: {tx['user_id']}, "
                  f"Amount: ${tx['amount']:.2f}, Status: {tx['status']}, Type: {tx['transaction_type']}")
            if tx['fraud_flags']:
                print(f"    Fraud Flags: {', '.join(tx['fraud_flags'])}")
            if tx.get('manual_review_flags'):
                print(f"    Manual Review: {tx['manual_review_flags'][-1]['reason']}")


        print("\n-- Transactions with Fraud Flags --")
        flagged_txns = self.filter_transactions(has_fraud_flags=True)
        if not flagged_txns:
            print("  No transactions with automatic fraud flags.")
        for tx in flagged_txns[:3]: # Display first 3
            print(f"  ID: {tx['transaction_id']}, User: {tx['user_id']}, Amount: ${tx['amount']:.2f}, Flags: {', '.join(tx['fraud_flags'])}")
            self.flag_transaction_for_review(tx['transaction_id'], "Automatic high value flag", "system_auto_flag")


        print("\n-- Filter Example: Failed transactions for user007 --")
        user_failed_txns = self.filter_transactions(status="failed", user_id="user007")
        if not user_failed_txns:
            print("  No matching transactions found for user007 with status failed.")
        for tx in user_failed_txns:
            print(f"  ID: {tx['transaction_id']}, Time: {tx['timestamp'].strftime('%Y-%m-%d %H:%M')}, Amount: ${tx['amount']:.2f}")

        print("------------------------------------")

if __name__ == '__main__':
    screen = TransactionMonitoringScreen()
    screen.display_screen()
